substrate/client/src/light.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Light client backend. Only stores headers and justifications of blocks.
+//! Everything else is requested from full nodes on demand.
+
+use std::sync::Arc;
+use primitives;
+use primitives::block::{self, Id as BlockId, HeaderHash};
+use state_machine::CodeExecutor;
+use state_machine::backend::Backend as StateBackend;
+use blockchain::{self, BlockStatus};
+use backend;
+use call_executor::RemoteCallExecutor;
+use client::{Client, GenesisBuilder};
+use error;
+use in_mem::{header_hash, Blockchain as InMemBlockchain};
+
+/// Light client data fetcher.
+pub trait Fetcher: Send + Sync {
+	/// Fetch method execution proof.
+	fn execution_proof(&self, block: HeaderHash, method: &str, call_data: &[u8]) -> error::Result<(Vec<u8>, Vec<Vec<u8>>)>;
```

Comment:
should be a `Future`

Reply:
maybe `CallExecutor` should return a future too. This will make it compose nicely with asynchronous RPC servers

Reply:
Changed `Fetcher` to return `Future`. Left `CallExecutor` as is for now